### PR TITLE
[2.9.0] Document the SQLite-backed caching experimental feature

### PR DESCRIPTION
**Pull request by moio**
Relates to https://github.com/rancher/rancher/issues/45635

## Description

This document the new SQLite-backed caching experimental feature.

### CODE REVIEW

**Comment by moio** [comment]
@richard-cox I added you as a reviewer as I have the feeling I might have forgotten something.

This is of course to be revised as implementation progresses.

**Comment by martyav** [comment]
The parent issue doesn't have a milestone. Is this for v2.9?

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+title: SQLite-backed caching
+---
+
+<head>
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching"/>
+</head>
+
+:::caution
+Enabling the SQLite-backed caching of Kubernetes objects is a **highly experimental** feature, it is not yet officially supported in Rancher. It is not intended for any production use. SUSE customers should consult SUSE Support before activating this feature.
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:10]:
```suggestion
Caching Kubernetes objects with SQLite is not intended for use in production. This feature is considered highly experimental. SUSE customers should consult SUSE Support before activating this feature.
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+:::
+
+
+## About this Feature
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:14]:
```suggestion
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+---
+title: SQLite-backed caching
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:2]:
```suggestion
title: SQLite-Backed Caching
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching"/>
+</head>
+
+:::caution
+Enabling the SQLite-backed caching of Kubernetes objects is a **highly experimental** feature, it is not yet officially supported in Rancher. It is not intended for any production use. SUSE customers should consult SUSE Support before activating this feature.
+:::
+
+
+## About this Feature
+This feature enables an optional cache of Kubernetes objects with the aim of improving performance, especially while browsing tabular data in the UI with thousands of rows. It also provides improved UI sorting, filtering and pagination features.
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:15]:
```suggestion
SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance, especially while browsing tabular data with thousands of rows. The cache improves UI sorting, filtering, and pagination. 

This feature creates file system based caches in the `rancher` pods of the upstream cluster, and in the `cattle-cluster-agent` pods of the downstream clusters. In most environments, disk usage and I/O should not be significant. However, you should monitor activity after you enable caching.

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. See [Encrypting SQLite-backed Caching](#encrypting-sqlite-backed-caches) if this is a security concern.
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+## Enabling instructions
+To enable the feature, perform the following steps:
+1. Log into the Rancher UI.
+1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
+1. Go to **`ui-sql-cache`** and click **⋮ > Activate > Activate**.
+1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
+1. In the upper left corner, click **☰ > Global Settings > Performance**.
+1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.
+1. Click **Apply**
+1. Refresh the page (`CTRL + R`)
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:26]:
```suggestion
## Enabling SQLite-Backed Caching

1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
1. Find **`ui-sql-cache`** and select **⋮ > Activate > Activate**.
1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
1. In the upper left corner, click **☰ > Global Settings > Performance**.
1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.
1. Click **Apply**.
1. Refresh the page (`CTRL + R`).
```

Is there no button to refresh the page? If not, different OS's have different combinations of key presses to perform a refresh and I guess we'd have to list them out

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+## Operational notes
+This feature creates file system based caches in the `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in downstream caches. In most use cases disk usage and I/O should not be significant, however monitoring is recommended.
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:29]:
```suggestion
```

Moved into intro and removed header

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+## Security notes
+This feature persists copies of any cached Kubernetes objects to disk. If this is a security concern, an option exists to have all objects encrypted before being persisted to disk.
+
+Set the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in the pods that need full encryption (`rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in downstream caches respectively).
+
+Secrets and security Tokens are always encrypted regardless of the above setting.
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:36]:
```suggestion
## Encrypting SQLite-backed Caches

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects  before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.

Secrets and security Tokens are always encrypted regardless of the above setting.
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+## Known limitations
+At the time of writing the feature is subject to limitations below:
+
+1. Only the Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.
+1. Automatic refresh of pages is disabled. Table contents can be manually refreshed via a **Refresh** button.
```
Comment (martyav) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:42]:
```suggestion
## Known Limitations of SQLite-Backed Caching

Only Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.

Pages can't be automatically refreshed. You can manually refresh table contents by clicking the **Refresh** button.
```

**Review by martyav — CHANGES_REQUESTED** [review]
Similar changes should be applied to the v2.9 file

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md (hunk 5 above)
Comment (moio) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md]: I addressed the refresh hotkey question in afe50d3d2a5

**Comment by moio** [comment]
> The parent issue doesn't have a milestone. Is this for v2.9?

Yes. I added the milestone to this PR - parent issue is all for 2.9, but not all for 2.9.0 (part of that will be in subsequent patch releases)

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+SQLite-backed caching persists copies of any cached Kubernetes objects to disk. See [Encrypting SQLite-backed Caching](#encrypting-sqlite-backed-caches) if this is a security concern.
+
+## Enabling SQLite-Backed Caching
+
+1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
+1. Find **`ui-sql-cache`** and select **⋮ > Activate > Activate**.
+1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
+1. In the upper left corner, click **☰ > Global Settings > Performance**.
+1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.
```
Comment (richard-cox) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:26]:
```suggestion
1. Go to **Server-side Pagination** and check the **Enable Server-side Pagination** option.
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+<head>
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching"/>
+</head>
+
+:::caution
+Caching Kubernetes objects with SQLite is not intended for use in production. This feature is considered highly experimental. SUSE customers should consult SUSE Support before activating this feature.
+:::
+
+
+SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance, especially while browsing tabular data with thousands of rows. The cache improves UI sorting, filtering, and pagination.
```
Comment (richard-cox) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:14]:
```suggestion
SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance. This unlocks sorting, filtering, and pagination features used by the UI to restrict the amount of resources it fetches and stores locally. These features are primarily used to improve list performance for resources with high counts by utilizing server-side pagination.
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+## Encrypting SQLite-backed Caches
+
+SQLite-backed caching persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects  before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.
+
+Secrets and security Tokens are always encrypted regardless of the above setting.
+
+## Known Limitations of SQLite-Backed Caching
+
+Only Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.
```
Comment (richard-cox) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:39]:
```suggestion
This initial release improves the performance of Pods, Secrets, Nodes and ConfigMaps in the Cluster Explorer pages, and most resources in the Explorer's `More Resources` section
```

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
+---
+title: SQLite-Backed Caching
```
Comment (richard-cox) [inline docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:2]:
```suggestion
title: SQLite-Backed Caching and UI Server-Side Pagination
```

Wonder if we need to make this a little clearer i.e backend cache changes and frontend changes that make use of it. I've tweaked a few things below, but could we call this out in the title?

**Review by richard-cox — CHANGES_REQUESTED** [review]

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md (hunk 12 above)
Comment (moio) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:2]: From a user perspective I actually think only UI is relevant - use of SQLite and fact that pagination is computed "Server-side" are implementation details.

From an engineering perspective, we are changing the cache implementation to enable better pagination. Both aspects are equally important and have to go hand in hand. I do not care which one is given more prominence.

How about just using

```suggestion
title: UI Server-Side Pagination
```

And clarifying later that is achieved via a new SQLite cache?

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md (hunk 11 above)
Comment (richard-cox) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md]: This second part of this list covers the other resources that make use of the cache

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md (hunk 12 above)
Comment (richard-cox) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md:2]: Sounds good!

**Comment by moio** [comment]
@richard-cox all feedback addressed, please take a look and approve if OK

@martyav I am not sure about the meaning of the `MERGE ON RELEASE` label - FMPOV it can be merged as soon as @richard-cox gives the green light as the feature has been present since 2.9.0 alpha5 few weeks ago.

Thanks everybody for the cooperation!

**Comment by martyav** [comment]
@moio thanks for clarifying. If the label is not necessary, I can remove. On our site, merging leads to immediate publication, so we usually hold PRs related to upcoming releases until the official release date. Though, in the case of 2.9 and other minor versions, merged 2.9.0 pages will appear in the 2.9 preview section of the docs site

**Review by richard-cox — APPROVED** [review]

**Comment by moio** [comment]
@martyav I confirm this is OK to merge anytime from our perspective.

**Review by martyav — CHANGES_REQUESTED** [review]
Mostly capitalization

**Comment by martyav** [comment]
@moio I think we also need to mention the flag in https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/feature-flags. I realize this is pretty late to bring up, but today I've been working on some other PRs related to that page. We can either add the update for https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/feature-flags here or I can open a second PR to address it.

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
+---
+title: UI Server-Side Pagination
+---
+
+<head>
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination"/>
+</head>
+
+:::caution
+UI Server-Side Pagination is not intended for use in production at this time. This feature is considered highly experimental. SUSE customers should consult SUSE Support before activating this feature.
```
[the remark this answers is not in the document]
Comment (moio) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md:10]: I thought SUSE Support is the official capitalization, as seen for example in

https://www.suse.com/support/

Are you sure we want to keep this lowercase?

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+## Encrypting SQLite-backed Caches
+
+UI Server-Side Pagination persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.
+
+Secrets and security Tokens are always encrypted regardless of the above setting.
+
+## Known Limitations of UI Server-Side Pagination
+
+This initial release improves the performance of Pods, Secrets, Nodes and ConfigMaps in the Cluster Explorer pages, and most resources in the Explorer's `More Resources` section.
```
[the remark this answers is not in the document]
Comment (moio) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md:39]: It was my understanding we always capitalize Kubernetes resources - eg ConfigMap in 

https://ranchermanager.docs.rancher.com/integrations-in-rancher/cis-scans/custom-benchmark#1-prepare-the-custom-benchmark-version-configmap

(and very many other examples in our docs, as well as Kubernetes upstream: https://kubernetes.io/docs/concepts/workloads/ mentioned Pods, ReplicaSets...)

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+1. Go to **Server-side Pagination** and check the **Enable Server-side Pagination** option.
+1. Click **Apply**.
+1. Reload the page with the browser button (or the equivalent keyboard combination, typically `CTRL + R` on Windows and Linux, and `⌘ + R` on macOS).
+
+
+## Encrypting SQLite-backed Caches
+
+UI Server-Side Pagination persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.
+
+Secrets and security Tokens are always encrypted regardless of the above setting.
```
[the remark this answers is not in the document]
Comment (moio) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md:35]: Same question in https://github.com/rancher/rancher-docs/pull/1337/files#r1698063308

**Comment by moio** [comment]
> Mostly capitalization, changes should also be applied to v2.9 page

Done. I did not change capitalization of SUSE Support and Kubernetes resources names yet per questions.

> I think we also need to mention the flag in [ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/feature-flags](https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/feature-flags).

Done.

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (martyav) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md]: Oops, then ignore! Thank you for pointing that out!

File: docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (martyav) [reply docs/how-to-guides/advanced-user-guides/enable-experimental-features/ui-server-side-pagination.md]: Sorry for overgeneralizing, and thank you for pushing back. We should be good now.

**Pull request merged**